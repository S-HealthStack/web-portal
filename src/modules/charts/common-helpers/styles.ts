import styled, { css } from 'styled-components';

import { colors, typography, px, animation } from 'src/styles';
import { LINE_CLASS_NAME } from '../common-components/Line';
import { AREA_CLASS_NAME } from '../common-components/Area';
import { DOT_CLASS_NAME } from '../common-components/Dot';

interface ContainerProps {
  width: number;
  height: number;
}

export const DOTS_CONTAINER_CLASS_NAME = 'dots';
export const X_AXIS_CLASS_NAME = 'xAxis';
export const Y_AXIS_CLASS_NAME = 'yAxis';
export const ZOOM_CLASS_NAME = 'zoom';

export const Container = styled.div.attrs<ContainerProps>(({ width, height }) => ({
  style: {
    width: px(width),
    height: px(height),
    position: 'relative',
  },
}))<ContainerProps>``;

export const TooltipContent = styled.div`
  display: flex;
  align-items: center;
`;

export const TooltipNameBlock = styled.div`
  display: flex;
  margin-right: ${px(8)};

  p {
    ${typography.labelSemibold};
    color: ${colors.surface};
    margin: 0;
  }
`;

export const TooltipDataBlock = styled.div`
  display: flex;
  flex-direction: column;

  p {
    ${typography.labelRegular};
    color: ${colors.surface};
    margin: 0;
  }
`;

export const TooltipColorPoint = styled.div<{ color: string }>`
  width: ${px(12)};
  height: ${px(12)};
  background-color: ${({ color }) => color};
  border-radius: 50%;
  margin-right: ${px(8)};
`;

export const brushOverlayStyles = css`
  fill: ${colors.primary};
  fill-opacity: 0.1;
  pointer-events: none;
`;

export const brushSelectionStyles = css`
  fill: ${colors.primary};
  fill-opacity: 1;
`;

export const brushHandleCommonStyles = css`
  stroke-width: ${px(1)};
  stroke: ${colors.primary};
  stroke-opacity: 0.3;
  fill: ${colors.onPrimary};
`;

export const StyledSvg = styled.svg<{
  focusWidth: number;
  $contextVisible: boolean;
}>`
  .${X_AXIS_CLASS_NAME} {
    .tick line {
      opacity: 0;
    }

    .tick {
      :first-child text {
        text-anchor: start;
      }
      :last-child text {
        text-anchor: end;
      }
    }
  }

  .${Y_AXIS_CLASS_NAME} {
    .tick line {
      stroke: ${colors.updBackgoundLight};
    }
  }

  .context {
    opacity: ${({ $contextVisible }) => ($contextVisible ? 1 : 0)};
    transition: opacity 0.3s ${animation.defaultTiming};
    .brushX {
      .overlay {
        ${brushOverlayStyles};
        height: ${px(4)};
      }
      .selection {
        ${brushSelectionStyles};
        height: ${px(22)};
        fill: url(#horizontalGradient);
      }
      .handle {
        ${brushHandleCommonStyles};
        height: ${px(22)};
        width: ${px(6)};
      }
    }
    .brushY {
      .overlay {
        ${brushOverlayStyles};
        width: ${px(4)};
      }
      .selection {
        ${brushSelectionStyles};
        width: ${px(22)};
        fill: url(#verticalGradient);
      }
      .handle {
        ${brushHandleCommonStyles};
        height: ${px(6)};
        width: ${px(22)};
      }
    }
  }

  .focus {
    cursor: move;
  }

  .${LINE_CLASS_NAME} {
    clip-path: url(#clip);
    transition: stroke-opacity 0.3s ${animation.defaultTiming};
  }

  .${AREA_CLASS_NAME} {
    clip-path: url(#clip);
    transition: fill-opacity 0.3s ${animation.defaultTiming};
  }

  .${DOTS_CONTAINER_CLASS_NAME} {
    clip-path: url(#clip);
    .${DOT_CLASS_NAME} {
      transition: fill-opacity 0.3s ${animation.defaultTiming};
    }
  }

  .zoom {
    fill: none;
    pointer-events: all;
  }
`;
