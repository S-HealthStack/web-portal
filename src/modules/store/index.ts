export type { RootState, AppDispatch, AppThunk } from './store';
export { useAppDispatch, useAppSelector } from './hooks';
export * from './tools';
