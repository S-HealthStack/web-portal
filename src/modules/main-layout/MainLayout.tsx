import React, { RefObject, useCallback, useEffect, useLayoutEffect, useRef } from 'react';
import { matchPath, Redirect, Route, Switch, useHistory, useLocation } from 'react-router-dom';
import useEvent from 'react-use/lib/useEvent';
import useMount from 'react-use/lib/useMount';
import useToggle from 'react-use/lib/useToggle';
import usePrevious from 'react-use/lib/usePrevious';
import styled from 'styled-components';
import { Location } from 'history';

import { userRoleSelector } from 'src/modules/auth/auth.slice';
import { UserRole } from 'src/modules/auth/userRole';
import { Path } from 'src/modules/navigation/store';
import StudySettings from 'src/modules/study-settings/StudySettings';
import Overview from 'src/modules/overview/Overview';
import OverviewSubject from 'src/modules/overview/overview-subject/OverviewSubject';
import SurveyEditor from 'src/modules/trial-management/survey-editor/SurveyEditor';
import SurveyPage from 'src/modules/trial-management/SurveyPage';
import DataInsights from 'src/modules/data-collection/DataInsights';
import { SnackbarContainer } from 'src/modules/snackbar';
import { useAppDispatch, useAppSelector } from 'src/modules/store';
import { fetchStudies, useSelectedStudyId } from 'src/modules/studies/studies.slice';
import { hideSnackbar } from 'src/modules/snackbar/snackbar.slice';
import { colors } from 'src/styles';
import { scrollToTop } from 'src/common/utils/scrollToTop';
import useDisableElasticScroll from 'src/common/useDisableElasticScroll';
import StudyManagement from 'src/modules/trial-management/StudyManagement';
import { LayoutContentCtx } from 'src/modules/main-layout/LayoutContentCtx';

import Sidebar from './sidebar/Sidebar';
import SwitchStudy from './switch-study/SwitchStudy';
import EmptyTab from './EmptyTab';

export const Layout = styled.div<{ isSwitchStudy?: boolean }>`
  width: 100%;
  height: 100%;
  padding: 0;
  transition: 1.4s cubic-bezier(0.45, 0.05, 0, 1);
  transform: translateY(${({ isSwitchStudy }) => (isSwitchStudy ? 0 : '-100%')});
`;

export const ContentWrapper = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  background-color: ${colors.background};
`;

export const Content = styled.div`
  min-height: unset;
  flex: 1;
  overflow: auto;
  position: relative;
`;

const useSwitchStudy = () => {
  const layoutRef = useRef<HTMLDivElement>(null);
  const [isSwitchStudy, toggleIsSwitchStudy] = useToggle(false);
  const [isSwitchStudyInTransition, toggleIsSwitchStudyInTransition] = useToggle(false);

  const isSelfElement = useCallback((evt: Event) => evt.currentTarget === evt.target, []);

  const onTransitionStart = useCallback(
    (evt: Event) => {
      if (isSelfElement(evt)) {
        toggleIsSwitchStudyInTransition(true);
      }
    },
    [isSelfElement, toggleIsSwitchStudyInTransition]
  );

  useEvent('transitionstart', onTransitionStart, layoutRef.current);

  const onTransitionEnd = useCallback(
    (evt: Event) => {
      if (isSelfElement(evt)) {
        toggleIsSwitchStudyInTransition(false);
      }
    },
    [isSelfElement, toggleIsSwitchStudyInTransition]
  );

  useEvent('transitionend', onTransitionEnd, layoutRef.current);

  return {
    layoutRef,
    isSwitchStudy,
    isSwitchStudyInTransition,
    toggleIsSwitchStudy,
  };
};

interface UseFetchBootDataReturnType {
  userRole?: UserRole;
}

const useFetchBootData = (): UseFetchBootDataReturnType => {
  const userRole = useAppSelector(userRoleSelector);
  const dispatch = useAppDispatch();

  useMount(() => {
    dispatch(fetchStudies());
  });

  return { userRole };
};

const useScrollHistory = (paths: Path[], content: RefObject<HTMLElement>) => {
  const pathMap = useRef<{ [key: string]: number }>({});
  const location = useLocation();
  const history = useHistory();
  const prevLocation = usePrevious<Location>(location);
  const pathHistory = useRef<string[]>([location.pathname]);

  useLayoutEffect(() => {
    if (prevLocation?.pathname === location.pathname) {
      return;
    }

    if (history.action === 'PUSH') {
      const prevPath = pathHistory.current[pathHistory.current.length - 1];

      paths.forEach((path) => {
        const matched = matchPath(prevPath, { path });

        if (matched && matched.isExact) {
          pathMap.current[path] = content.current?.scrollTop || 0;
        }
      });

      if (content.current) {
        content.current.scrollTop = 0;
      }

      pathHistory.current.push(location.pathname);
    } else {
      if (pathHistory.current.length < 2) {
        return;
      }
      pathHistory.current.pop();

      const currentPath = pathHistory.current[pathHistory.current.length - 1];

      paths.forEach((path) => {
        const matched = matchPath(currentPath, { path });

        if (matched && matched.isExact && content.current) {
          content.current.scrollTop = pathMap.current[path];
          delete pathMap.current[path];
        }
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [location]);
};

const MainLayout = () => {
  const contentRef = useRef<HTMLDivElement>(null);
  const location = useLocation();
  const studyId = useSelectedStudyId();
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(hideSnackbar());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [location.pathname, studyId]);

  useDisableElasticScroll(contentRef);

  useScrollHistory([Path.Overview, Path.DataCollection, Path.TrialManagement], contentRef);

  const { userRole } = useFetchBootData();

  const { layoutRef, isSwitchStudy, toggleIsSwitchStudy, isSwitchStudyInTransition } =
    useSwitchStudy();

  const onStudyClick = () => {
    if (isSwitchStudy) {
      toggleIsSwitchStudy();
    } else {
      scrollToTop(contentRef.current as HTMLElement, toggleIsSwitchStudy);
    }
  };

  return (
    <LayoutContentCtx.Provider value={contentRef}>
      <Layout ref={layoutRef} isSwitchStudy={isSwitchStudy}>
        <ContentWrapper>
          <SwitchStudy onStudySelectionFinished={toggleIsSwitchStudy} />
        </ContentWrapper>
        <ContentWrapper>
          <Sidebar onStudyClick={onStudyClick} />
          {userRole && (
            <Content ref={contentRef}>
              <Switch>
                <Route exact path={Path.Overview} component={Overview} />
                <Route exact path={Path.OverviewSubject} component={OverviewSubject} />
                <Route exact path={Path.TrialManagement} component={StudyManagement} />
                <Route exact path={Path.TrialManagementEditSurvey} component={SurveyEditor} />
                <Route exact path={Path.TrialManagementSubject} component={OverviewSubject} />
                <Route exact path={Path.TrialManagementSurveyResults} component={SurveyPage} />
                <Route path={Path.UserAnalytics} component={EmptyTab} />
                <Route exact path={Path.DataCollection} component={DataInsights} />
                <Route exact path={Path.DataCollectionSubject} component={OverviewSubject} />
                <Route exact path={Path.StudySettings}>
                  <StudySettings
                    isSwitchStudy={isSwitchStudy}
                    isSwitchStudyInTransition={isSwitchStudyInTransition}
                  />
                </Route>
                <Redirect to={Path.Overview} />
              </Switch>
              <SnackbarContainer useSimpleGrid />
            </Content>
          )}
        </ContentWrapper>
      </Layout>
    </LayoutContentCtx.Provider>
  );
};

export default MainLayout;
