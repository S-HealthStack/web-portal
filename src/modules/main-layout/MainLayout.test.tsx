import React from 'react';
import 'jest-styled-components';
import 'src/__mocks__/setupWindowMatchMediaMock';
import 'src/__mocks__/setupResizeObserverMock';
import 'src/__mocks__/setupRangeMock';

import { ThemeProvider } from 'styled-components/';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import { store } from 'src/modules/store/store';
import { theme } from 'src/styles';
import { act, render, screen } from '@testing-library/react';
import { history } from 'src/modules/navigation/store';
import MainLayout from 'src/modules/main-layout/MainLayout';

describe('MainLayout', () => {
  it('should render', async () => {
    await act(() => {
      render(
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <ConnectedRouter history={history}>
              <MainLayout />
            </ConnectedRouter>
          </Provider>
        </ThemeProvider>
      );
    });

    expect(await screen.findByTestId('main-layout')).toMatchSnapshot();
  });
});
