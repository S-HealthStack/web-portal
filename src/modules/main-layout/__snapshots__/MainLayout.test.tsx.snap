// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout should render 1`] = `
.c10 {
  position: absolute;
  will-change: opacity,transform;
  border-radius: 50%;
  -webkit-animation: bykygw 600ms cubic-bezier(0.4,0,0.2,1) forwards;
  animation: bykygw 600ms cubic-bezier(0.4,0,0.2,1) forwards;
  z-index: 0;
  background-color: #E3EAFB;
  pointer-events: none;
}

.c7 {
  position: relative;
  display: -webkit-box !important;
  display: -webkit-flex !important;
  display: -ms-flexbox !important;
  display: flex !important;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.25rem;
  width: 5rem;
  white-space: nowrap;
  padding: 0;
  overflow: hidden;
  -webkit-transition-duration: 0;
  transition-duration: 0;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  height: 3rem;
  border: none;
  background-color: transparent;
}

.c7 > div {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 600;
  font-style: normal;
}

.c7 > div svg {
  margin-right: 0.25rem;
}

.c7 > div {
  color: #4475E3;
}

.c7 > div svg {
  fill: #4475E3;
}

.c7:hover:enabled > div {
  color: #2F65E0;
}

.c7:hover:enabled > div svg {
  fill: #2F65E0;
}

.c7:hover:enabled:active > div {
  color: #2057D5;
}

.c7:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c7:disabled > div {
  color: #B3C6F1;
}

.c7:disabled > div svg {
  fill: #B3C6F1;
}

.c7:focus-visible {
  outline: 0.125rem solid #C7D6F7;
}

.c7:focus-visible > div {
  color: #4475E3;
}

.c7:focus-visible > div svg {
  fill: #4475E3;
}

.c25 {
  position: relative;
  display: -webkit-box !important;
  display: -webkit-flex !important;
  display: -ms-flexbox !important;
  display: flex !important;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.25rem;
  width: 100%;
  white-space: nowrap;
  padding: 0;
  overflow: hidden;
  -webkit-transition-duration: 0;
  transition-duration: 0;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  height: 1.5rem;
  width: 1.5rem;
  border: none;
  background-color: transparent;
}

.c25:hover {
  cursor: pointer;
}

.c25 > div {
  color: #4475E3;
}

.c25 > div svg {
  fill: #4475E3;
}

.c25:hover:enabled > div {
  color: #2F65E0;
}

.c25:hover:enabled > div svg {
  fill: #2F65E0;
}

.c25:hover:enabled:active > div {
  color: #2057D5;
}

.c25:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c25:disabled > div {
  color: #B3C6F1;
}

.c25:disabled > div svg {
  fill: #B3C6F1;
}

.c25:focus-visible {
  outline: 0.125rem solid #C7D6F7;
}

.c25:focus-visible > div {
  color: #4475E3;
}

.c25:focus-visible > div svg {
  fill: #4475E3;
}

.c25:hover:enabled {
  background-color: #ECF1FC;
}

.c25:hover:enabled:active > div {
  color: #2057D5;
}

.c25:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c25:focus-visible:enabled {
  background-color: #C7D6F7;
  outline: none;
}

.c9 {
  position: relative;
  z-index: 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transform: translate3d(0,0,0.0625rem);
  -ms-transform: translate3d(0,0,0.0625rem);
  transform: translate3d(0,0,0.0625rem);
}

.c24 {
  display: inline-block;
  position: static;
}

.c29 {
  border-radius: 1.5625rem;
  height: 2.5rem;
  -webkit-transition: opacity 300ms cubic-bezier(0.45,0.05,0,1);
  transition: opacity 300ms cubic-bezier(0.45,0.05,0,1);
  position: relative;
  background-color: transparent !important;
  opacity: 1;
}

.c29:hover {
  opacity: 1;
  box-shadow: 0 0.9375rem 1.25rem rgba(180,180,180,0.3);
  cursor: pointer;
}

.c29:hover:active {
  opacity: 1;
  box-shadow: 0 0.9375rem 1.25rem rgba(180,180,180,0.3);
  cursor: pointer;
  -webkit-filter: drop-shadow(0 0.9375rem 1.25rem rgba(180,180,180,0.3));
  filter: drop-shadow(0 0.9375rem 1.25rem rgba(180,180,180,0.3));
}

.c19 {
  border-radius: 4.125rem;
  height: 8.875rem;
  -webkit-transition: opacity 300ms cubic-bezier(0.45,0.05,0,1);
  transition: opacity 300ms cubic-bezier(0.45,0.05,0,1);
  position: relative;
  background-color: transparent !important;
  opacity: 1;
}

.c19:hover {
  opacity: 1;
  box-shadow: 0 0.9375rem 1.25rem rgba(180,180,180,0.3);
  cursor: pointer;
}

.c19:hover:active {
  opacity: 1;
  box-shadow: 0 0.9375rem 1.25rem rgba(180,180,180,0.3);
  cursor: pointer;
  -webkit-filter: drop-shadow(0 0.9375rem 1.25rem rgba(180,180,180,0.3));
  filter: drop-shadow(0 0.9375rem 1.25rem rgba(180,180,180,0.3));
}

.c23 {
  height: 100%;
  background-color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #808080;
  position: relative;
  border-right: 0.0625rem solid #F7F8FA;
  -webkit-transition: border 150ms cubic-bezier(0.45,0.05,0,1);
  transition: border 150ms cubic-bezier(0.45,0.05,0,1);
  padding-bottom: 0.875rem;
}

.c27 {
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  opacity: 1;
}

.c41 {
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  opacity: 0;
}

.c31 {
  margin-right: 1rem;
}

.c26 {
  position: absolute !important;
  z-index: 1100;
  background-color: #F6F8FE;
  top: -62.5rem;
  left: -62.5rem;
}

.c28 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem;
  margin-bottom: 2.375rem;
}

.c28:hover {
  cursor: pointer;
}

.c28:hover .c30 {
  box-shadow: 0 0.5rem 0.625rem 0.125rem 0 rgba(0,0,0,0.12);
}

.c32 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  margin-top: 0.125rem;
  word-break: break-word;
  display: block;
  color: #474747;
}

.c42 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding-left: 1rem;
  position: relative;
  overflow: hidden;
  height: 3.9375rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 0.25rem;
  font-family: Inter,sans-serif;
  font-size: 0.75rem;
  line-height: 1.125rem;
  font-weight: 400;
  font-style: normal;
  color: #2D2D2D !important;
}

.c42 svg {
  fill: #808080;
}

.c43 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  margin-bottom: 0.5rem;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 100%;
}

.c34 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  overflow: auto;
  position: relative;
}

.c38 {
  position: relative;
  z-index: 1000;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 0.5rem;
}

.c40 {
  position: relative;
  z-index: 1000;
  margin: 0;
  display: block;
}

.c36 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  overflow: hidden;
  z-index: 0;
}

.c37 {
  padding: 1rem;
  height: 2.5rem;
  margin-bottom: 0.5rem;
  color: #808080 !important;
}

.c37 svg {
  fill: #808080;
}

.c37:hover {
  cursor: pointer;
  background-color: #ECF1FC;
  color: #4475E3 !important;
}

.c37:hover svg {
  fill: #4475E3;
}

.c37::before {
  content: '';
  height: 2rem;
  width: 0.25rem;
  border-radius: 0.125rem 0rem 0rem 0.125rem;
  position: absolute;
  left: 11.75rem;
  background-color: #4475E3;
  color: #4475E3;
  z-index: 100;
  display: none;
}

.c37 > .c39 {
  font-family: Inter,sans-serif;
  font-size: 0.75rem;
  line-height: 1.125rem;
  font-weight: 400;
  font-style: normal;
}

.c35 {
  margin-top: auto;
}

.c35 .c33 {
  opacity: 0;
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c35:hover .c33 {
  opacity: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  height: 100%;
  width: 100%;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 {
  width: 100%;
  height: 100%;
  background-color: #F7F8FA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: auto;
}

.c8 {
  margin-top: 5.9375rem;
  height: 0;
}

.c11 {
  min-width: 70rem;
  max-width: 70rem;
  color: #474747;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-top: 2.125rem;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 4.25rem;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c15 {
  min-width: 70rem;
  max-width: 70rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow-x: hidden;
  margin-bottom: 1.875rem;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  white-space: nowrap;
  -webkit-transform: translateX(0%);
  -ms-transform: translateX(0%);
  transform: translateX(0%);
  -webkit-transition: -webkit-transform 0.3s cubic-bezier(0.45,0.05,0,1);
  -webkit-transition: transform 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: transform 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c17 {
  min-width: 70rem;
  max-width: 70rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: auto;
  min-height: 27.25rem;
}

.c13 {
  font-family: Inter,sans-serif;
  font-size: 1.5rem;
  line-height: 2.0125rem;
  font-weight: 600;
  font-style: normal;
  margin: 0.3125rem 0 0 2.5rem;
}

.c18 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 9rem;
  height: 13.125rem;
  padding-top: 0.25rem;
  margin: 0 2.5rem 0.5rem;
}

.c20 {
  font-family: Inter,sans-serif;
  font-size: 1.125rem;
  line-height: 1.6875rem;
  font-weight: 400;
  font-style: normal;
  margin-top: 1.625rem;
  text-align: center;
  max-width: 9rem;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c21 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #4475E3;
}

.c5 {
  margin-bottom: 5.75rem;
}

.c0 {
  width: 100%;
  height: 100%;
  padding: 0;
  -webkit-transition: 1.4s cubic-bezier(0.45,0.05,0,1);
  transition: 1.4s cubic-bezier(0.45,0.05,0,1);
  -webkit-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  transform: translateY(-100%);
}

.c1 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #F7F8FA;
}

.c2 {
  overflow: auto;
}

.c22 {
  position: relative;
  z-index: 1;
}

<div
  class="c0"
  data-testid="main-layout"
>
  <div
    class="c1 c2"
  >
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <button
              class="c7 c8"
              disabled=""
              fill="text"
              style="overflow: hidden; position: relative;"
              width="80"
            >
              <div
                class="c9"
                data-testid="content"
              >
                <svg />
              </div>
              <div
                class="c10"
                color="primaryLightPressed"
                opacity="1,1"
                scale="0"
                style="display: none; width: 0rem; height: 0rem; top: 0rem; left: 0rem;"
              />
            </button>
            <div
              class="c11"
            >
              <div
                class="c12"
              >
                <div
                  class="c13"
                >
                  Study Collection
                </div>
              </div>
              <div
                class="c14"
              >
                <div
                  class="c15"
                >
                  <div
                    class="c16"
                  >
                    <div
                      class="c17"
                    >
                      <div
                        class="c18"
                      >
                        <div
                          class="c19"
                          color="secondarySkyBlue"
                        >
                          <svg
                            data-testid="avatar-icon"
                            fill="#00B0D7"
                          />
                        </div>
                        <div
                          class="c20"
                        >
                          SleepCare Study
                        </div>
                      </div>
                      <div
                        class="c18"
                      >
                        <div
                          class="c19"
                          color="secondaryViolet"
                        >
                          <svg
                            data-testid="avatar-icon"
                            fill="#944ED7"
                          />
                        </div>
                        <div
                          class="c20"
                        >
                          Heart Health Study
                        </div>
                      </div>
                      <div
                        class="c18"
                      >
                        <div
                          class="c19"
                          color="disabled"
                        >
                          <svg
                            data-testid="avatar-icon"
                            fill="#DBDBDB"
                          />
                        </div>
                        <div
                          class="c20"
                        >
                          Some deleted study
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="c21"
              >
                 
              </div>
            </div>
            <button
              class="c7 c8"
              disabled=""
              fill="text"
              style="overflow: hidden; position: relative;"
              width="80"
            >
              <div
                class="c9"
                data-testid="content"
              >
                <svg />
              </div>
              <div
                class="c10"
                color="primaryLightPressed"
                opacity="1,1"
                scale="0"
                style="display: none; width: 0rem; height: 0rem; top: 0rem; left: 0rem;"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c1 c22"
  >
    <div
      class="c23"
      id="sidebar"
      style="width: 12rem; padding-top: 1.625rem;"
    >
      <span
        class="c24"
        data-testid="tooltip-container"
      />
      <button
        class="c25 c26"
        fill="text"
        style="overflow: hidden; position: relative;"
      >
        <div
          class="c9"
          data-testid="content"
        >
          <svg />
        </div>
        <div
          class="c10"
          color="primaryLightPressed"
          opacity="1,1"
          scale="0"
          style="display: none; width: 0rem; height: 0rem; top: 0rem; left: 0rem;"
        />
      </button>
      <div
        class="c27 c28"
      >
        <div
          class="c29 c30 c31"
          color="secondarySkyBlue"
        >
          <svg
            data-testid="avatar-icon"
            fill="#00B0D7"
          />
        </div>
        <div
          class="c32"
        >
          SleepCare Study
        </div>
      </div>
      <div
        class="c33 c34"
      />
      <div
        class="c35"
      >
        <div
          class="c33 c34"
        >
          <div
            class="c36 c37"
            data-testid="user-profile-signout-action"
          >
            <div
              class="c38"
            >
              <svg />
            </div>
            <p
              class="c39 c40"
            >
              Sign out
            </p>
            <div
              class="c10"
              color="primaryLightPressed"
              opacity="1,1"
              scale="0"
              style="display: none; width: 0rem; height: 0rem; top: 0rem; left: 0rem;"
            />
          </div>
        </div>
        <div
          class="c41 c42"
        >
          <div
            class="c43"
          />
          Unknown
        </div>
      </div>
    </div>
  </div>
</div>
`;
